In [ ]:
# Data Loading and cleanup
import pandas as pd
import numpy as nm

df = pd.read_csv("src/data.csv")
df.shape

(541909, 8)

In [3]:
# first look
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

Transactional data by orders/invoice. Head shows single customer with one order with multiple items. StockCodes appear to have a non-uniform formatting. 

In [4]:
#Data Types
df.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

1. optimize datatypes  
InvoiceNo => int? why object?  
InvoiceDate => DateTime  
CustomerID => int
2. Add TotalPrice
3. Check for negative quantities
4. Check for negative prices or 0
5. Missing CustomerID
6. (opt. add weekday for analysis?)


In [5]:
#convert InvoiceNo to string
df["InvoiceNo"] = df["InvoiceNo"].astype(str)
# Check
df["InvoiceNo"].str.isdigit().value_counts()

InvoiceNo
True     532618
False      9291
Name: count, dtype: int64

9291 rows with non-numerical values

In [ ]:
#preview non-numerical values
df[~df["InvoiceNo"].str.isdigit()]["InvoiceNo"].unique()

array(['C536379', 'C536383', 'C536391', ..., 'C581499', 'C581568',
       'C581569'], dtype=object)

In [ ]:
#Extract non-numerical values
cancelled = df.copy()
cancelled["OrderStatus"] = cancelled["InvoiceNo"].str.extract(r'^([A-Za-z])')
cancelled["OrderStatus"].value_counts()

OrderStatus
C    9288
A       3
Name: count, dtype: int64

C Values represent canceled invoices  
add column "OrderStatus"  
A Values as completed

In [ ]:
#add column "OrderStatus"
cancelled["OrderStatus"] = cancelled["InvoiceNo"].apply(lambda x: "cancelled" if x.startswith('C') else "completed")
cancelled

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country OrderStatus  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom   completed  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom   completed  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom   completed  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom   completed  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom   completed  
...                 ...        ...         ...             ...         ...  
541904  12/9/2011 12:50       0.85     12680.0          France   completed  
541905  12/9/2011 12:50       2.10     12680.0          France   completed  
541906  12/9/2011 12:50       4.15     12680.0          France   completed  
541907  12/9/2011 12:50       4.15     12680.0          France   completed  
541908  12/9/2011 12:50       4.95     12680.0          France   completed  

[541909 rows x 9 columns]

In [13]:
#Searching for missing values
data.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [15]:
#Calculating impact of missing CustomerID data
missingID = data["CustomerID"].isnull().sum() / data.shape[0] * 100
missingID.round(2)

24.93

Out of the 541909 records there are 1454 with missing descriptions. Also a large number of missing CustomerIDs. ~24.93%

In [21]:
#convert InvoiceDate to DateTime, CustomerID to Int64 because of missing values
cancelled["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])
cancelled["CustomerID"] = cancelled["CustomerID"].astype("Int64")
cancelled


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country OrderStatus  
0      2010-12-01 08:26:00       2.55       17850  United Kingdom   completed  
1      2010-12-01 08:26:00       3.39       17850  United Kingdom   completed  
2      2010-12-01 08:26:00       2.75       17850  United Kingdom   completed  
3      2010-12-01 08:26:00       3.39       17850  United Kingdom   completed  
4      2010-12-01 08:26:00       3.39       17850  United Kingdom   completed  
...                    ...        ...         ...             ...         ...  
541904 2011-12-09 12:50:00       0.85       12680          France   completed  
541905 2011-12-09 12:50:00       2.10       12680          France   completed  
541906 2011-12-09 12:50:00       4.15       12680          France   completed  
541907 2011-12-09 12:50:00       4.15       12680          France   completed  
541908 2011-12-09 12:50:00       4.95       12680          France   completed  

[541909 rows x 9 columns]

In [ ]:
#add TotalPrice column for total price per item
totalprice = cancelled.copy()
totalprice["TotalPrice"] = totalprice["Quantity"] * totalprice ["UnitPrice"]
totalprice

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country OrderStatus  \
0      2010-12-01 08:26:00       2.55       17850  United Kingdom   completed   
1      2010-12-01 08:26:00       3.39       17850  United Kingdom   completed   
2      2010-12-01 08:26:00       2.75       17850  United Kingdom   completed   
3      2010-12-01 08:26:00       3.39       17850  United Kingdom   completed   
4      2010-12-01 08:26:00       3.39       17850  United Kingdom   completed   
...                    ...        ...         ...             ...         ...   
541904 2011-12-09 12:50:00       0.85       12680          France   completed   
541905 2011-12-09 12:50:00       2.10       12680          France   completed   
541906 2011-12-09 12:50:00       4.15       12680          France   completed   
541907 2011-12-09 12:50:00       4.15       12680          France   completed   
541908 2011-12-09 12:50:00       4.95       12680          France   completed   

        TotalPrice  
0            15.30  
1            20.34  
2            22.00  
3            20.34  
4            20.34  
...            ...  
541904       10.20  
541905       12.60  
541906       16.60  
541907       16.60  
541908       14.85  

[541909 rows x 10 columns]

In [30]:
#Check for negative quantities
(totalprice["Quantity"] < 0).any()
(totalprice["Quantity"] < 0).value_counts()
totalprice[totalprice["Quantity"] < 0].head()


InvoiceNo StockCode                       Description  Quantity  \
141   C536379         D                          Discount        -1   
154   C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235   C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236   C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237   C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   

            InvoiceDate  UnitPrice  CustomerID         Country OrderStatus  \
141 2010-12-01 09:41:00      27.50       14527  United Kingdom   cancelled   
154 2010-12-01 09:49:00       4.65       15311  United Kingdom   cancelled   
235 2010-12-01 10:24:00       1.65       17548  United Kingdom   cancelled   
236 2010-12-01 10:24:00       0.29       17548  United Kingdom   cancelled   
237 2010-12-01 10:24:00       0.29       17548  United Kingdom   cancelled   

     TotalPrice  
141      -27.50  
154       -4.65  
235      -19.80  
236       -6.96  
237       -6.96

In [53]:
totalprice.shape

(541909, 10)

negative quanties (10624) not equal to cancelled orders (9288)

In [39]:
neg_qty_completed = totalprice[
    (totalprice["Quantity"] < 0) &
    (totalprice["OrderStatus"] == "completed")
    ]
neg_qty_completed

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
2406      536589     21777         NaN       -10 2010-12-01 16:50:00   
4347      536764    84952C         NaN       -38 2010-12-02 14:42:00   
7188      536996     22712         NaN       -20 2010-12-03 15:30:00   
7189      536997     22028         NaN       -20 2010-12-03 15:30:00   
7190      536998     85067         NaN        -6 2010-12-03 15:30:00   
...          ...       ...         ...       ...                 ...   
535333    581210     23395       check       -26 2011-12-07 18:36:00   
535335    581212     22578        lost     -1050 2011-12-07 18:38:00   
535336    581213     22576       check       -30 2011-12-07 18:38:00   
536908    581226     23090     missing      -338 2011-12-08 09:56:00   
538919    581422     23169     smashed      -235 2011-12-08 15:24:00   

        UnitPrice  CustomerID         Country OrderStatus  TotalPrice  
2406          0.0        <NA>  United Kingdom   completed        -0.0  
4347          0.0        <NA>  United Kingdom   completed        -0.0  
7188          0.0        <NA>  United Kingdom   completed        -0.0  
7189          0.0        <NA>  United Kingdom   completed        -0.0  
7190          0.0        <NA>  United Kingdom   completed        -0.0  
...           ...         ...             ...         ...         ...  
535333        0.0        <NA>  United Kingdom   completed        -0.0  
535335        0.0        <NA>  United Kingdom   completed        -0.0  
535336        0.0        <NA>  United Kingdom   completed        -0.0  
536908        0.0        <NA>  United Kingdom   completed        -0.0  
538919        0.0        <NA>  United Kingdom   completed        -0.0  

[1336 rows x 10 columns]

In [34]:
neg_qty_completed["Description"].unique()

array([nan, '?', 'check', 'damages', 'faulty', 'Dotcom sales',
       'reverse 21/5/10 adjustment', 'mouldy, thrown away.', 'counted',
       'Given away', 'Dotcom', 'label mix up', 'samples/damages',
       'thrown away', 'incorrectly made-thrown away.', 'showroom', 'MIA',
       'Dotcom set', 'wrongly sold as sets', 'Amazon sold sets',
       'dotcom sold sets', 'wrongly sold sets', '? sold as sets?',
       '?sold as sets?', 'Thrown away.', 'damages/display',
       'damaged stock', 'broken', 'throw away', 'wrong barcode (22467)',
       'wrong barcode', 'barcode problem', '?lost',
       "thrown away-can't sell.", "thrown away-can't sell", 'damages?',
       're dotcom quick fix.', "Dotcom sold in 6's", 'sold in set?',
       'cracked', 'sold as 22467', 'Damaged',
       'mystery! Only ever imported 1800',
       'MERCHANT CHANDLER CREDIT ERROR, STO', 'POSSIBLE DAMAGES OR LOST?',
       'damaged', 'DAMAGED', 'Display', 'Missing', 'wrong code?',
       'wrong code', 'adjust', 'crush

In [ ]:
#check for unrealistic prizes
neg_prc = totalprice[(totalprice["UnitPrice"] <= 0)]
neg_prc

InvoiceNo StockCode                   Description  Quantity  \
622       536414     22139                           NaN        56   
1970      536545     21134                           NaN         1   
1971      536546     22145                           NaN         1   
1972      536547     37509                           NaN         1   
1987      536549    85226A                           NaN         1   
...          ...       ...                           ...       ...   
536981    581234     72817                           NaN        27   
538504    581406    46000M  POLYESTER FILLER PAD 45x45cm       240   
538505    581406    46000S  POLYESTER FILLER PAD 40x40cm       300   
538554    581408     85175                           NaN        20   
538919    581422     23169                       smashed      -235   

               InvoiceDate  UnitPrice  CustomerID         Country OrderStatus  \
622    2010-12-01 11:52:00        0.0        <NA>  United Kingdom   completed   
1970   2010-12-01 14:32:00        0.0        <NA>  United Kingdom   completed   
1971   2010-12-01 14:33:00        0.0        <NA>  United Kingdom   completed   
1972   2010-12-01 14:33:00        0.0        <NA>  United Kingdom   completed   
1987   2010-12-01 14:34:00        0.0        <NA>  United Kingdom   completed   
...                    ...        ...         ...             ...         ...   
536981 2011-12-08 10:33:00        0.0        <NA>  United Kingdom   completed   
538504 2011-12-08 13:58:00        0.0        <NA>  United Kingdom   completed   
538505 2011-12-08 13:58:00        0.0        <NA>  United Kingdom   completed   
538554 2011-12-08 14:06:00        0.0        <NA>  United Kingdom   completed   
538919 2011-12-08 15:24:00        0.0        <NA>  United Kingdom   completed   

        TotalPrice  
622            0.0  
1970           0.0  
1971           0.0  
1972           0.0  
1987           0.0  
...            ...  
536981         0.0  
538504         0.0  
538505         0.0  
538554         0.0  
538919        -0.0  

[2517 rows x 10 columns]

In [51]:
#check for negative prizes
neg_prc = totalprice[(totalprice["UnitPrice"] < 0)]
neg_prc

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299983   A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
299984   A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

        UnitPrice  CustomerID         Country OrderStatus  TotalPrice  
299983  -11062.06        <NA>  United Kingdom   completed   -11062.06  
299984  -11062.06        <NA>  United Kingdom   completed   -11062.06

cancel reasons are listed as items, missing descriptions, "bad debt"  
remove rows to avoid bias in sales statistics

In [67]:
#cleaned up data
df_sales = totalprice.drop(neg_prc.index)
df_sales

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country OrderStatus  \
0      2010-12-01 08:26:00       2.55       17850  United Kingdom   completed   
1      2010-12-01 08:26:00       3.39       17850  United Kingdom   completed   
2      2010-12-01 08:26:00       2.75       17850  United Kingdom   completed   
3      2010-12-01 08:26:00       3.39       17850  United Kingdom   completed   
4      2010-12-01 08:26:00       3.39       17850  United Kingdom   completed   
...                    ...        ...         ...             ...         ...   
541904 2011-12-09 12:50:00       0.85       12680          France   completed   
541905 2011-12-09 12:50:00       2.10       12680          France   completed   
541906 2011-12-09 12:50:00       4.15       12680          France   completed   
541907 2011-12-09 12:50:00       4.15       12680          France   completed   
541908 2011-12-09 12:50:00       4.95       12680          France   completed   

        TotalPrice  
0            15.30  
1            20.34  
2            22.00  
3            20.34  
4            20.34  
...            ...  
541904       10.20  
541905       12.60  
541906       16.60  
541907       16.60  
541908       14.85  

[539392 rows x 10 columns]

In [72]:
df_sales.to_csv("output/sales_data.csv", sep=";")

In [ ]:
#Check NA CustomerID
df_sales[df_sales["CustomerID"].isna()]

InvoiceNo StockCode                      Description  Quantity  \
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
1447      536544     21790               VINTAGE SNAP CARDS         9   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

               InvoiceDate  UnitPrice  CustomerID         Country OrderStatus  \
1443   2010-12-01 14:32:00       2.51        <NA>  United Kingdom   completed   
1444   2010-12-01 14:32:00       2.51        <NA>  United Kingdom   completed   
1445   2010-12-01 14:32:00       0.85        <NA>  United Kingdom   completed   
1446   2010-12-01 14:32:00       1.66        <NA>  United Kingdom   completed   
1447   2010-12-01 14:32:00       1.66        <NA>  United Kingdom   completed   
...                    ...        ...         ...             ...         ...   
541536 2011-12-09 10:26:00       4.13        <NA>  United Kingdom   completed   
541537 2011-12-09 10:26:00       4.13        <NA>  United Kingdom   completed   
541538 2011-12-09 10:26:00       4.96        <NA>  United Kingdom   completed   
541539 2011-12-09 10:26:00      10.79        <NA>  United Kingdom   completed   
541540 2011-12-09 10:26:00    1714.17        <NA>  United Kingdom   completed   

        TotalPrice  
1443          2.51  
1444          5.02  
1445          3.40  
1446          3.32  
1447         14.94  
...            ...  
541536       20.65  
541537       16.52  
541538        4.96  
541539       10.79  
541540     1714.17  

[132603 rows x 10 columns]